python/dgl/dataloading/negative_sampler.py

```diff
@@ -65,3 +66,64 @@ def _generate(self, g, eids, canonical_etype):
         src = F.repeat(src, self.k, 0)
         dst = F.randint(shape, dtype, ctx, 0, g.number_of_nodes(vtype))
         return src, dst
+
+# Alias
+Uniform = PerSourceUniform
+
+class GlobalUniform(_BaseNegativeSampler):
+    """Negative sampler that randomly chooses negative source-destination pairs according
+    to a uniform distribution.
+
+    For each edge ``(u, v)`` of type ``(srctype, etype, dsttype)``, DGL generates
+    :attr:`k` pairs of negative edges ``(u', v')``, where ``u'`` is chosen uniformly from
+    all the nodes of type ``srctype`` and ``v'`` is chosen uniformly from all the nodes
+    of type ``dsttype``.  The resulting edges will also have type
+    ``(srctype, etype, dsttype)``.
+
+    Parameters
+    ----------
+    k : int
+        The number of negative examples to generate per minibatch.
+    exclude_self_loops : bool, optional
+        Whether to exclude self-loops from negative examples.  (Default: False)
+    unique : bool, optional
+        Whether to sample unique negative examples.  Setting it to False will make things
+        faster.  (Default: True)
+    num_trials : int, optional
+        The number of rejection sampling trials.
+
+        Increasing it will increase the likelihood of getting :attr:`k` negative examples,
+        but will also take more time if the graph is dense.
+    redundancy : float, optional
+        Indicates how much more negative examples to actually generate during rejection sampling
+        before finding the unique pairs.
+
+        Only effective if :attr:`unique` is True.
+
+        Increasing it will increase the likelihood of getting :attr:`k` negative examples,
+        but will also take more time and memory.
+
+    Notes
+    -----
+    This negative sampler may not always return :attr:`k` negative samples.  Consider
```

Comment:
Might be worth a brief explanation as to why this might happen--e.g., a small graph generates many redundant edges when random sampling and/or the graph is so dense that not many unique negative edges exist, but that this shouldn't be the case for medium sized graphs that are relatively sparse.

Reply:
Good point.